tests/avify4mtest.c

```diff
@@ -0,0 +1,177 @@
+// Copyright 2022 Google LLC. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "apps/shared/y4m.h"
+#include "avif/avif.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static const char kPath[] = "/tmp/avify4mtest.y4m";
+
+//------------------------------------------------------------------------------
+
+// Returns true if image1 and image2 are identical.
+static avifBool compareYUVA(const avifImage * image1, const avifImage * image2)
+{
+    if (image1 == NULL || image2 == NULL || image1->width != image2->width || image1->height != image2->height ||
+        image1->depth != image2->depth || image1->yuvFormat != image2->yuvFormat || image1->yuvRange != image2->yuvRange) {
+        printf("ERROR: input mismatch\n");
+        return AVIF_FALSE;
+    }
+
+    avifPixelFormatInfo formatInfo;
+    avifGetPixelFormatInfo(image1->yuvFormat, &formatInfo);
+    uint32_t uvWidth = (image1->width + formatInfo.chromaShiftX) >> formatInfo.chromaShiftX;
+    uint32_t uvHeight = (image1->height + formatInfo.chromaShiftY) >> formatInfo.chromaShiftY;
+    if (uvWidth < 1)
+        uvWidth = 1;
+    if (uvHeight < 1)
+        uvHeight = 1;
+
+    for (uint32_t plane = 0; plane < (formatInfo.monochrome ? 1 : AVIF_PLANE_COUNT_YUV); ++plane) {
+        const uint32_t pixelByteCount =
+            ((plane == AVIF_CHAN_Y) ? image1->width : uvWidth) * ((image1->depth > 8) ? sizeof(uint16_t) : sizeof(uint8_t));
+        const uint32_t height = (plane == AVIF_CHAN_Y) ? image1->height : uvHeight;
+        for (uint32_t y = 0; y < height; ++y) {
+            if (memcmp(image1->yuvPlanes[plane] + y * image1->yuvRowBytes[plane],
```

Comment:
Since `image1->yuvRowBytes[plane]` is a uint32_t:

    uint32_t yuvRowBytes[AVIF_PLANE_COUNT_YUV];

the product `y * image1->yuvRowBytes[plane]` can potentially overflow a uint32_t **in general**. In this test, I am sure it won't overflow, but it would be better to eliminate this concern. A common solution is to cast one of the numbers to `size_t`. Alternatively you can declare and increment a row pointer, as is done in y4m.c.